### ISSUE TITLE
XGBoost hyper tuned parameters give a lower accuracy compared to the default model

### ISSUE THREAD

**Issue by nelepi** [nelepi/xgbooost_questions#1]
Hi all! 

Problem: XGBoost hyper tuned parameters give a lower accuracy compared to the default model.

I have a question regarding the XGBoost algorithm. The goal is to predict a multiclass classification variable. The dataframe consists of 2600 observations with 55 features. Im working in R and therefore performing hyperparameter tuning with the MLR package. The test and train data set are stratified with 20-80% division. The data is a bit imbalanced.

test_y
  0   1   2   3 
278  64  37 155 

train_y
   0    1    2    3 
1126  216  159  637 

As stated above, the problem is that the accuracy of the default model is lower (for both the test and training dataset). The code below. Does anyone know where I am making a mistake? Or what can I check or improve? Code below.

The evaluation metric of the xgboost training 'mlogloss' and of the mlr training is 'mmce'. Is that a problem?

Thanks in advance!! Im a bit new to ml algorithms, so all help is very welcome.

`### multiclass classification: mc; all1
#R: https://www.hackerearth.com/practice/machine-learning/machine-learning-algorithms/beginners-tutorial-on-xgboost-parameter-tuning-r/tutorial/

### parameters
num_class = 4
boost = "gbtree"
obj = "multi:softprob"
eval_m = "mlogloss"
watchlist <- list(train = xgb_train, test = xgb_test)
seed = seed

### CV with default values
params <- list(booster = boost, 
               objective = obj, 
               eta=0.1, 
               gamma=0, 
               max_depth=6, 
               min_child_weight=1, 
               subsample=0.8, 
               colsample_bytree=0.8,
               #alpha = 0,
               #lambda = 1,
               num_class = num_class)

xgbcv_def <- fun_xgbcv(params)  

# evaluation
min_iter_xgbcv_def <- fun_eval_xgbcv(xgbcv_def)

### train with nrounds from CV 
xgb_train1 <- fun_xgbtrain(params, min_iter_xgbcv_def)

# evaluation on train data
evalxgb_train1_train <- fun_eval_xgbtrain(xgb_train1, xgb_train, train_y)
conftable_xgb_train1_train <- evalxgb_train1_train[1] 
accuracy_xgb_train1_train <- evalxgb_train1_train[2]  

# evaluation on test data
evalxgb_train1_test <- fun_eval_xgbtrain(xgb_train1, xgb_test, test_y)
conftable_xgb_train1_test <- evalxgb_train1_test[1]
accuracy_xgb_train1_test <- evalxgb_train1_test[2]

### hyper parameter tuning with MLR package
#getParamSet("classif.xgboost")
traintask <- makeClassifTask(data = df_train, 
                             target = "more_than_X_perc_damage")
testtask <- makeClassifTask(data = df_test, 
                            target = "more_than_X_perc_damage")

# learner
lrn <- makeLearner("classif.xgboost",
                   predict.type = "prob", #"response"
                   par.vals = list(objective=obj, 
                                   eval_metric=eval_m,
                                   nrounds= 100))

# set resampling strategy (cross-validation)
rdesc <- makeResampleDesc("CV",stratify = T,iters=10L) #10L ***

# set random search in param set
ctrl <- makeTuneControlRandom(maxit = 1000L) #1000L ***

# parameter space ***
params_to_tune <- makeParamSet(makeIntegerParam("max_depth",lower =3L,upper = 10L),
                       makeNumericParam("min_child_weight",lower = 0L,upper = 5L),
                       makeNumericParam("subsample",lower = 0.5,upper = 1),
                       makeNumericParam("eta", lower = 0.03, upper = 0.3),
                       makeNumericParam("colsample_bytree",lower = 0.5,upper = 1),
                       makeNumericParam("gamma",lower = 0,upper = 100),
                       makeNumericParam("alpha",lower = -1, upper = 0, trafo = function(x) 10^x),
                       makeNumericParam("lambda", lower = -1, upper = 0, trafo = function(x) 10^x))

# set parallel backend
#parallelStartSocket(cpus = detectCores())

# tune parameters
tuned_params <- tuneParams(learner = lrn,
                           task = traintask, 
                           resampling = rdesc,
                           measures = mmce, #multiclass.brier
                           par.set = params_to_tune, 
                           control = ctrl,
                           show.info = T)

#parallelStop()
#mmce_mean = tuned_params$y
#print(mmce_mean)
min_multiclass_brier_test_mean= tuned_params$y
print(min_multiclass_brier_test_mean)

# new model with tuned pars
xgb_tuned_learner <- setHyperPars(learner = lrn,
                                  par.vals = tuned_params$x)

# retrain model with tuned pars
xgb_model <- mlr::train(xgb_tuned_learner, traintask)

# prediction performance
pred <- predict(xgb_model, testtask)
performance(pred) 
calculateConfusionMatrix(pred)
confusionMatrix(pred$data$response,pred$data$truth)

# optimal parameters 
params_tuned <- list(booster = boost,
                             objective = obj,
                             eta = tuned_params$x$eta,
                             gamma = tuned_params$x$gamma,
                             max_depth =  tuned_params$x$max_depth,
                             min_child_weight = tuned_params$x$min_child_weight,
                             subsample = tuned_params$x$subsample,
                             colsample_bytree = tuned_params$x$colsample_bytree,
                             alpha = tuned_params$x$alpha,
                             lambda = tuned_params$x$lambda,
                             num_class = num_class)

### CV for min_iter with tuned params
xgb_cv2 <- fun_xgbcv(params_tuned)

# evaluation
min_iter_xgbcv2 <- fun_eval_xgbcv(xgb_cv2)

### train with nrounds from CV 
xgb_train2 <- fun_xgbtrain(params_tuned, min_iter_xgbcv2)

# evaluation on train data
evalxgb_train2_train <- fun_eval_xgbtrain(xgb_train2, xgb_train, train_y)
conftable_xgb_train2_train <- evalxgb_train2_train[1]
accuracy_xgb_train2_train <- evalxgb_train2_train[2]

# evaluation on test data
evalxgb_train2_test <- fun_eval_xgbtrain(xgb_train2, xgb_test, test_y)
conftable_xgb_train2_test <- evalxgb_train2_test[1]
accuracy_xgb_train2_test <- evalxgb_train2_test[2]

### comparison
print("TRAIN")
print("Accuracy:")
print("default:")
print(accuracy_xgb_train1_train)
print("tuned:")
print(accuracy_xgb_train2_train)

print("Table:")
print("default:")
print(conftable_xgb_train1_train)
print("tuned:")
print(conftable_xgb_train2_train)

print("TEST")
print("Accuracy:")
print("default:")
print(accuracy_xgb_train1_test)
print("tuned:")
print(accuracy_xgb_train2_test)

print("Table:")
print("default:")
print(conftable_xgb_train1_test)
print("tuned:")
print(conftable_xgb_train2_test)

# weird that tuned model performs worse in comparison to default model
# something with weights; imbalance, or missing values

### comparison parameters
print("eta (d-t):")
params$eta
params_tuned$eta

print("gamma (d-t):")
params$gamma
params_tuned$gamma

print("max_depth (d-t):")
params$max_depth
params_tuned$max_depth

print("min_child_weight (d-t):")
params$min_child_weight
params_tuned$min_child_weight

print("subsample (d-t):")
params$subsample
params_tuned$subsample

print("colsample_bytree (d-t):")
params$colsample_bytree
params_tuned$colsample_bytree

print("alpha (d-t):")
0
params_tuned$alpha

print("lambda (d-t):")
1
params_tuned$lambda

### save
#df_params_tuned = as.data.frame(params_tuned)
#write.csv(df_params_tuned,paste0(main_directory, "models/df_tuned_params1.csv"), row.names = FALSE)`

`### multiclass classification: functions

## xgb cv
fun_xgbcv <- function(params){
  xgbcv <- xgb.cv(params = params, 
                  data = xgb_train, 
                  nrounds = 300, 
                  nfold = 10, 
                  showsd = T, 
                  stratified = T, 
                  print_every_n = 10, 
                  early_stop_round = 20, 
                  maximize = F,
                  metrics = eval_m)
  return(xgbcv)
}

## xgb train
fun_xgbtrain <- function(params, nrounds){
  xgbtrain <- xgb.train(params = params, 
                        data = xgb_train, 
                        nrounds = nrounds, 
                        watchlist = watchlist, 
                        print_every_n = 10, 
                        early_stop_round = 20, 
                        maximize = F , 
                        eval_metric = eval_m)
  return(xgbtrain)
}

## evaluate xgb cv
fun_eval_xgbcv <- function(model){
  d <- as.data.frame(model$evaluation_log)
  
  # plot
  print(ggplot(d) + 
    geom_line(aes(x = iter, y = train_mlogloss_mean, col = "train")) +
    geom_line(aes(x = iter, y = test_mlogloss_mean, col = "test")) +
    scale_y_continuous(sec.axis = sec_axis(~.*1, name="test_mlogloss")) +
    theme(legend.position="right") +
    ggtitle(deparse(substitute(model))))
  
  # best iteration
  min_iter = d[d$test_mlogloss_mean == min(d$test_mlogloss_mean),]$iter
  return(min_iter)
}

## evaluate xgb train
fun_eval_xgbtrain <- function(model, newdata, label){
  d <- as.data.frame(model$evaluation_log)
  
  # plot
  print(ggplot(d) + 
    geom_line(aes(x = iter, y = train_mlogloss, col = "train")) +
    geom_line(aes(x = iter, y = test_mlogloss, col = "test")) +
    scale_y_continuous(sec.axis = sec_axis(~.*1, name="test_mlogloss")) +
    theme(legend.position="right")+
    ggtitle(deparse(substitute(model))))
  
  # importance
  #imp <- xgb.importance(colnames(xgb_train), model=model)
  #print(imp)
  #print(xgb.plot.importance(imp)) #using gain
  
  # prediction
  pred <- predict(model, newdata =newdata) #pred_def[1]+pred_def[2]+pred_def[3]+pred_def[4] =1
  pred_mat <- matrix(pred, nrow = num_class, ncol= length(pred)/num_class) %>%
    t() %>%
    data.frame() %>%
    mutate(label=label, max_prob = max.col(., "last")-1) 
  conf_table = table(Prediction=pred_mat$max_prob, Actual = pred_mat$label)
  accuracy = sum(diag(conf_table))/nrow(newdata)
  
  return(list(conf_table, accuracy))
}`
